### PR TITLE
EVG-14957 Add linter for ParserProject and the merge functions

### PR DESCRIPTION
**Pull request by bynn**
[EVG-14957](https://jira.mongodb.org/browse/EVG-14957)

### Description 
add shell script to check for changes in parserproject struct and the merge functions 

### Testing 
verified with local testing by making a new branch and manually setting the common ancestor.

### CODE REVIEW

**Review by ablack12 — CHANGES_REQUESTED** [review]
Awesome! Going to let Kim review the script but looks good to me.

We'll want to also add this task to the list of [Github patch definition tasks](https://evergreen.mongodb.com/projects##evergreen). (We could add a "custom-linter" tag to verify-agent-version-update and this new task, and then just modify the patch definition to use the "custom-linter" tag instead of the agent regex.

File: scripts/verify-merge-function-update.sh [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
+
```
Comment (Kimchelly) [inline scripts/verify-merge-function-update.sh:2]: I would add a comment here describing what this script is intended to do, because it's difficult to understand its purpose just from reading it.

**Review by Kimchelly — CHANGES_REQUESTED** [review]
I feel like this shell script is kind of fragile because it will produce false results. For example, it can't handle changes unrelated to adding/removing fields from the ParserProject struct, so if you do something like add an irrelevant code change like a comment or a struct tag, this script will still error. I would add to the top-level comment on the script warning it's not fool-proof.

This script also has a lot more edge cases than the agent version update script. Did you test that this works locally and in CI when the ParserProject is changed and when it's unchanged?

File: scripts/verify-merge-function-update.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+common_ancestor=$(git merge-base ${BRANCH_NAME}@{upstream} HEAD);
+files_changed="$(git diff --name-only "${common_ancestor}" -- 'model/project_parser_merge_functions.go')"
+if [[ "${files_changed}" == "" ]]; then
+    merge_function_changed=true
+fi
+
+# Find the latest commit when the ParserProject struct was updated.
+project_parser_start_line_num=$(git grep -n -e "type ParserProject struct" model/project_parser.go | cut -d ':' -f 2);
+project_parser_end_line_num=$(git grep -n -e "End of ParserProject struct" model/project_parser.go | cut -d ':' -f 2);
+commit_project_parser_main=$(git blame -l -L ${project_parser_start_line_num},${project_parser_end_line_num} main -- model/project_parser.go | cut -d ' ' -f 1);
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh:23]: done

File: scripts/verify-merge-function-update.sh (hunk 1 above)
Comment (bynn) [reply scripts/verify-merge-function-update.sh:3]: done

File: scripts/verify-merge-function-update.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+set -eo pipefail
+
+if [[ "${BRANCH_NAME}" == "" ]]; then
+    BRANCH_NAME=main;
+fi
+
+merge_function_changed=false
+parser_project_changed=false
+
+# Find the common ancestor between the current set of changes and the upstream branch, then see if any source code files
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: done

File: scripts/verify-merge-function-update.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# have changed in the merge functions.
+common_ancestor=$(git merge-base ${BRANCH_NAME}@{upstream} HEAD);
+files_changed="$(git diff --name-only "${common_ancestor}" -- 'model/project_parser_merge_functions.go')"
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: done

File: scripts/verify-merge-function-update.sh [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+if [[ "${files_changed}" == "" ]]; then
+    merge_function_changed=true
+fi
+
+# Find the latest commit when the ParserProject struct was updated.
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: done

File: scripts/verify-merge-function-update.sh [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+project_parser_start_line_num=$(git grep -n -e "type ParserProject struct" model/project_parser.go | cut -d ':' -f 2);
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: file path put into variable but the struct itself is called `ParserProject` but the file it's in is called `project_parser` so maybe we can make it consistent in other ticket.

File: scripts/verify-merge-function-update.sh [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+project_parser_end_line_num=$(git grep -n -e "End of ParserProject struct" model/project_parser.go | cut -d ':' -f 2);
+commit_project_parser_main=$(git blame -l -L ${project_parser_start_line_num},${project_parser_end_line_num} main -- model/project_parser.go | cut -d ' ' -f 1);
+last_commit_project_parser_updated=$(git blame -l -L ${project_parser_start_line_num},${project_parser_end_line_num} model/project_parser.go | cut -d ' ' -f 1);
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: done

File: scripts/verify-merge-function-update.sh [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+# Check if the last commit when the ParserProject struct was updated is in the list of committed changes.
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: done

File: scripts/verify-merge-function-update.sh [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        parser_project_changed=true
+    fi
+done
+
+if [[ "${merge_function_changed}" == "${parser_project_changed}" ]]; then
+    exit 0
+fi
+
+if ${merge_function_changed}; then 
+    echo "project_parser_merge_functions has been changed but ParserProject struct has not been updated. Please update project_parser_merge_functions." >&2
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: done

File: scripts/verify-merge-function-update.sh [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+elif ${parser_project_changed}; then
+    echo "ParserProject struct has been changed but project_parser_merge_functions has not been updated. Please update the ParserProject struct." >&2
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: done

File: scripts/verify-merge-function-update.sh [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    if [[ "${line}" == "0000000000000000000000000000000000000000" ]]; then
+        parser_project_changed=true
+    fi
+done
+
+if [[ "${merge_function_changed}" == "${parser_project_changed}" ]]; then
+    exit 0
+fi
+
+if ${merge_function_changed}; then
```
[the remark this answers is not in the document]
Comment (bynn) [reply scripts/verify-merge-function-update.sh]: script updated

File: scripts/verify-merge-function-update.sh [hunk 12]
```diff
@@ -1,48 +1,35 @@
 #!/usr/bin/env bash
+# Script makes sure that changes to the ParserProject struct and merge functions happen concurrently.
```
Comment (Kimchelly) [inline scripts/verify-merge-function-update.sh:2]: I would rephrase this slightly: "Script to verify that if the ParserProject struct is updated, its merge functions are also updated."

**Review by Kimchelly — COMMENTED** [review]
It mostly looks good, but I still have my question about the local/PR testing for the script.

File: model/project_parsor_merge_functions.go [hunk 13]
```diff
@@ -0,0 +1,275 @@
+package model
```
[the remark this answers is not in the document]
Comment (ablack12) [reply model/project_parsor_merge_functions.go:3]: can you move this to be after `package model` ?

File: scripts/verify-merge-function-update.sh [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    exit 0;
+fi
+
+# See if the project parser merge functions file has changed
+files_changed="$(git diff --name-only "${common_ancestor}" -- "${MERGE_FUNC_FILE_PATH}")"
+if [[ "${files_changed}" == "" ]]; then
+    exit 0;
+fi
+
+echo -e "ParserProject struct has been changed but ${MERGE_FUNC_FILE_PATH} has not been updated. Please update the ParserProject struct." >&2
```
Comment (ablack12) [inline scripts/verify-merge-function-update.sh:34]: "Please update the merge function" seems more like the request for developers

File: self-tests.yml [hunk 15]
```diff
@@ -593,6 +602,11 @@ tasks:
     commands:
       - func: get-project
       - func: verify-agent-version-update
+  - name: verify-merge-function-update
```
Comment (ablack12) [inline self-tests.yml:605]: I think we should add `tags: ["linter"]` here and to `verify-agent-version-update`. Then at line 655 instead of adding those tasks individually to the build variant we can say `-name: .linter` and Evergreen will include any tasks that have the linter tag. We can also use this to update the PR definitions on the project page. Let me know if you have a question about this.

File: model/project_parsor_merge_functions.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (bynn) [reply model/project_parsor_merge_functions.go:3]: done

File: scripts/verify-merge-function-update.sh (hunk 14 above)
Comment (bynn) [reply scripts/verify-merge-function-update.sh:34]: done

File: self-tests.yml (hunk 15 above)
Comment (bynn) [reply self-tests.yml:605]: done. 
 I wasn't going to update the project page until after it's merged so I might ask about it later!

**Review by bynn — COMMENTED** [review]
also made @Kimchelly's changs but can't seem to comment on it

**Review by ablack12 — APPROVED** [review]

**Comment by Kimchelly** [comment]
I still have the same question that I originally had from [this comment](https://github.com/evergreen-ci/evergreen/pull/5003#pullrequestreview-749504571).

**Comment by bynn** [comment]
I think it's because if you add a new field to the ParserProject struct, there wouldn't be a unit test that could test if that was merged in properly because the field hasn't existed before. The unit tests the current fields but I think this is just so that when something like that happens, it would be more obvious to the person making changes. 
@ablack12 Just to double check.

**Comment by Kimchelly** [comment]
Do the merge functions not get updated unit tests when changing fields? I feel like if this is an issue purely related to remembering changes in code, it should be standard to update unit tests when you add new code (which should also be caught by code review). Leaning on a shell script linters that use git to enforce code standards seems like an undesirable precedent.

**Comment by ablack12** [comment]
The idea behind the linter is because people would easily not know, when adding a new field to the parser project, that they need to add the new field to the merge function. This is hard to catch in a unit test because the unit test wouldn't know about the field yet. It's common for us to miss things like this -- Upserts that update each field individually for example have caused a lot of trouble in the past and we've ended up just refactoring them to not reference each field name. In this case I think the linter is a good idea and could save us a headache in the future.

If a field is renamed, this would also impact the merge function and so would still pass this test. A comment would fail the test but I think we could just read and understand what had happened in the test and still go forward, and comments aren't frequently changed without some field itself being impacted.

**Review by Kimchelly — APPROVED** [review]
This still seems like a fragile solution since it depends on the code being structured in a really specific way to satisfy the linter, which is even more non-obvious, complicated, and easily breakable than something like just putting a comment on the ParserProject struct saying "if you update this, update the merge functions as well". There's a lot of places in the code base where arguably, this sort of linter that remembers that "if you change this, also change that" (e.g. updating the REST models in sync with the DB models) should also apply, but it seems like using this shell script solution everywhere where we have to remember that kind of non-obvious update would be a bad idea.

But as long as the app team is fine with doing it like this, I guess it's fine. I would discourage doing this in general though. Code correctness should be enforced by unit tests and linters should be used for enforcing style and best practices. Instead of writing a shell script, I would have written a test using reflection to verify that you set all the fields fields and after running the project merger, all fields are non-zero (and if any are non-zero, you forgot to merge a field).

**Comment by ablack12** [comment]
I get what you're saying; not everything currently in the parser project struct needs to be added to merge though and reflection is fraught with its own drama. I think we can go ahead with this task and if it causes trouble or isn't catching things the way we would expect we can always revisit how we want to approach this.
